pkg/restore/change_storageclass_action.go

```diff
@@ -0,0 +1,104 @@
+limitations under the License.
+*/
+
+package restore
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	corev1apiclient "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/heptio/velero/pkg/plugin/framework"
+	"github.com/heptio/velero/pkg/plugin/velero"
+)
+
+// ChangeStorageClassAction updates a PV or PVC's storage class name
+// if a mapping is found in the plugin's config map.
+type ChangeStorageClassAction struct {
+	logger logrus.FieldLogger
+	client corev1apiclient.ConfigMapInterface
+}
+
+func NewChangeStorageClassAction(logger logrus.FieldLogger, client corev1apiclient.ConfigMapInterface) *ChangeStorageClassAction {
+	return &ChangeStorageClassAction{
+		logger: logger,
+		client: client,
+	}
+}
+
+func (a *ChangeStorageClassAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"persistentvolumeclaims", "persistentvolumes"},
+	}, nil
+}
+
+func (a *ChangeStorageClassAction) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
+	a.logger.Info("Executing ChangeStorageClassAction")
+	defer a.logger.Info("Done executing ChangeStorageClassAction")
+
+	a.logger.Debug("Getting plugin config")
+	config, err := getPluginConfig(framework.PluginKindRestoreItemAction, "velero.io/change-storageclass", a.client)
+	if err != nil {
+		return nil, err
+	}
+
+	if config == nil || len(config.Data) == 0 {
+		a.logger.Debug("No storage class mappings found")
+		return &velero.RestoreItemActionExecuteOutput{
+			UpdatedItem: input.Item,
+		}, nil
+	}
+
+	obj, ok := input.Item.(*unstructured.Unstructured)
+	if !ok {
+		return nil, errors.Errorf("object was of unexpected type %T", input.Item)
+	}
+
+	log := a.logger.WithFields(map[string]interface{}{
+		"kind":      obj.GetKind(),
+		"namespace": obj.GetNamespace(),
+		"name":      obj.GetName(),
+	})
+
+	// use the unstructured helpers here since this code is for both PVs and PVCs, and the
+	// field names are the same for both types.
+	storageClass, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "storageClassName")
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting item's spec.storageClassName")
+	}
+	if storageClass == "" {
+		log.Debug("Item has no storage class specified")
+		return &velero.RestoreItemActionExecuteOutput{
+			UpdatedItem: input.Item,
+		}, nil
+	}
+
+	newStorageClass, ok := config.Data[storageClass]
+	if !ok {
+		return &velero.RestoreItemActionExecuteOutput{
+			UpdatedItem: input.Item,
+		}, nil
+	}
+
+	log.Infof("Updating item's storage class name to %s", newStorageClass)
+
+	if err := unstructured.SetNestedField(obj.UnstructuredContent(), newStorageClass, "spec", "storageClassName"); err != nil {
+		return nil, errors.Wrap(err, "unable to set item's spec.storageClassName")
+	}
+
+	return velero.NewRestoreItemActionExecuteOutput(obj), nil
```

Comment:
should the above returns also be using `NewRestoreItemActionExecuteOutput` instead of directly creating the struct?